text/0000-app-mixins-stack-buildpacks.md

```diff
@@ -0,0 +1,335 @@
+1. During the lifecycle's build phase, the stackpacks that passed detection will run against the build and run images accordingly (see details below). All stackpacks will be run before the user's buildpacks.
+
+## Stack Buildpacks
+
+ A stack buildpacks (a.k.a. stackpacks) is a special case of buildpack that has the following properties:
+
+* Is run as the `root` user
+* Configured with `privileged = true` in the `buildpack.toml`
+* Can only create one layer
+* Does not have access to the application source code
+* Is run before all regular buildpacks
+* Is run against both the build and run images
+* Is distributed with the stack run and/or build image
+* May not write to the `/layers` directory
+
+The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.
+
+For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase excluding the a few specific directories and files.
+
+Alternatively, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+A stack can provide stackpacks by including them in the `/cnb/stack/buildpacks` directory. By default, an implicit order will be created for all stackpacks included in a stack. The order can be overridden in the `builder.toml` with the following configuration:
+
+```
+[[stack.order]]
+  [[stack.order.group]]
+    id = "<stackpack id>"
+```
+
+A stackpack will only execute if it passes detection. When the stackpack is executed, it's detect and build scripts use the same parameters as the regular buildpacks. But the arguments may differ. For example, the first positional argument to the `bin/build` (the `<layers>` directory) MUST NOT be a standard layers directory adhering to the [Buildpacks Build specification](https://github.com/buildpacks/spec/blob/main/buildpack.md#build).
+
+The stackpack's snapshot layer may be modified by writing a `stack.toml` file. The `stack.toml` will define paths that will be restored even when the base image changes (ex. package indicies) and paths that will be excluded from the launch image (ex. `/var/cache`).
+
+## Rebasing an App
+
+Before a launch image is rebased, the platform must re-run the any stackpacks that were used to build the launch image against the new run-image. Then, the rebase operation can be performed as normal, while including the stackpack layers as part of the stack. This will be made possible by including the stackpack in the run-image, but because the stackpack detect phase is not run, the operation does not need access to the application source.
```

Comment:
> the platform must re-run the any stackpacks that were used to build the launch image against the new run-image

Is this really a "must"? While this would produce that most comprehensive upgrade, a regular rebase should work without a full upgrade? I could see a regular rebase being preferable in some situations where performance is critical.

Reply:
I think this would introduce the risk of a stackpack appending to some file, but then the rebase also changes that files, and those rebase changes are lost.